src/cargo/util/rustc.rs

```diff
@@ -62,4 +75,157 @@ impl Rustc {
+                    Ok(serde_json::from_str(&json)?)
+                }
+            }
+            (_, fingerprint) => {
+                if let Err(e) = fingerprint {
+                    warn!("failed to calculate rustc fingerprint: {}", e);
+                }
+                info!("rustc info cache disabled");
+                Cache {
+                    cache_location: None,
+                    dirty: false,
+                    data: CacheData::default(),
+                }
+            }
+        }
+    }
+
+    fn cached_output(&mut self, cmd: &ProcessBuilder) -> CargoResult<(String, String)> {
+        let calculate = || {
+            let output = cmd.exec_with_output()?;
+            let stdout = String::from_utf8(output.stdout)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            let stderr = String::from_utf8(output.stderr)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            Ok((stdout, stderr))
+        };
+        if self.cache_location.is_none() {
+            info!("rustc info uncached");
+            return calculate();
+        }
+
+        let key = process_fingerprint(cmd);
+        match self.data.outputs.entry(key) {
+            Entry::Occupied(entry) => {
+                info!("rustc info cache hit");
+                Ok(entry.get().clone())
+            }
+            Entry::Vacant(entry) => {
+                info!("rustc info cache miss");
+                let output = calculate()?;
+                entry.insert(output.clone());
+                self.dirty = true;
+                Ok(output)
+            }
+        }
+    }
+}
+
+impl Drop for Cache {
+    fn drop(&mut self) {
+        match (&self.cache_location, self.dirty) {
+            (&Some(ref path), true) => {
+                let json = serde_json::to_string(&self.data).unwrap();
+                match paths::write(path, json.as_bytes()) {
+                    Ok(()) => info!("updated rustc info cache"),
+                    Err(e) => warn!("failed to update rustc info cache: {}", e),
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+fn rustc_fingerprint(path: &Path) -> CargoResult<u64> {
+    let mut hasher = SipHasher::new_with_keys(0, 0);
+
+    let path = paths::resolve_executable(path)?;
+    path.hash(&mut hasher);
+
+    paths::mtime(&path)?.hash(&mut hasher);
+
+    match (env::var("RUSTUP_HOME"), env::var("RUSTUP_TOOLCHAIN")) {
+        (Ok(rustup_home), Ok(rustup_toolchain)) => {
+            debug!("adding rustup info to rustc fingerprint");
+            rustup_toolchain.hash(&mut hasher);
+            rustup_home.hash(&mut hasher);
+            let rustup_rustc = Path::new(&rustup_home)
+                .join("toolchains")
+                .join(rustup_toolchain)
+                .join("bin")
+                .join("rustc");
+            paths::mtime(&rustup_rustc)?.hash(&mut hasher);
+        }
+        _ => (),
+    }
```

Comment:
This match requires some scrutiny during review, because it is not tested, but kinda important :-) 

We can't just look at `RUSTUP_TOOLCHAIN`, because the toolchain name does not change when you do `rustup update`, it is just `stable-x86_64-unknown-linux-gnu`. So we actually have to look inside the toolchain and mtime the real rustc binary. 

This still has a failure mode unfortunately :( If the `cargo` is not rustup, but `rustc` is rustup, then these env variables won't be set, and we won't be able to detect compieler changes.

Reply:
One possible mitigation here I think would be to look at the result of the resolved executable. If `rustc` resolves to `$CARGO_HOME/.bin/rustc` then I think we can say with a high level of certainty that it's a rustup rustc. We could even say that if we detect a rustup rustc and `RUSTUP_HOME` and/or `RUSTUP_TOOLCHAIN` isn't set that we skip this cache entirely.

Reply:
Yep, implemented that logic!